design_docs/2021-10_architecture_design_doc.md

```diff
@@ -0,0 +1,185 @@
+
+## Terms
+- “compiled layer” - the inner `Iter`-based layer which monomorphizes.
+    - “small operator” or just “operator” - An operator within the compiled layer.
+- “scheduled layer” - outer Timely-style scheduled layer.
+    - “compiled subtree” (or “fat operator”) - A unit of the scheduled graph containing a subgraph of small operators compiled and monomorphized together. What Timely would simply call an “operator”, to be confusing.
+
+In order to construct a Hydroflow graph, Hydroflow graphs must be partitioned into *compiled subtrees*. Each subgraph must be an *in-out tree* (or *bi-*[*arborescense*](https://en.wikipedia.org/wiki/Arborescence_(graph_theory))) which we define as a graph obeying the following constraints:
+
+1. It is a [directed tree](https://en.wikipedia.org/wiki/Polytree).
+2. The tree has a fan-in → fan-out structure and there is some “central vertex” which is on the path between every source/sink pair.
+    1. Formally, you can divide the vertices into an IN set, OUT set, and single ROOT vertex. From any vertex in the IN to any vertex in the OUT set there is exactly one path, and that path goes through the ROOT.
+
+Additionally for our compiled subtrees:
+
+3. Every edge pointing into such a subgraph points to a handoff.
+
+**Example #1** **(invalid)**
+
+![](images/2021-10_ADD_01.png)
+
+
+This compiled subgraph is invalid because the graph is not a directed tree.
+
+**Example #2** **(invalid)**
+
+![](images/2021-10_ADD_02.png)
+
+
+This compiled subgraph is invalid because the input edge does not point at a handoff.
+
+**Example #3** **(invalid)**
+
+![](images/2021-10_ADD_03.png)
+
+
+This compiled subgraph is invalid because there is no “central vertex.”
+
+These structural constraints ensure that that chunk of code will be easy to represent in straightforward (possibly machine-generated) Rust. This design is in service of providing a flexible mechanism for higher-level components to make policy decisions about where buffers are appropriate for their particular application. One way to view a Hydroflow graph is that it resembles a Timely Dataflow graph where each operator contains a chunk of code representing the compiled subgraph.
+
+## Example (valid)
+![](images/2021-10_ADD_04.png)
+
+
+There are of course multiple possible ways to partition this graph given the freedom to insert handoff points wherever we please.  In this diagram, `A`, `B`, `C`, and `D` would need handoff points which would accept input to be buffered for when their corresponding operators are scheduled. Another way to partition this would be `{A, B, O, C, D}` and `{P}` for example, which would require slightly different handoff points.
+
+The (monomorphized) code for the compiled subgraphs themselves might look something like this:
+
+**Subgraph** `**1**`
+
+    |a, b, c, d| {
```

Comment:
nit: consider switching these to use triple backticks (it's annoying that dropbox paper exports them like this)